### DISCUSSION TITLE
How to use matplotlib config for generated pages ?

### DISCUSSION THREAD

**Discussion by ksachdeva** [executablebooks/meta#151] [Q&A]
Hi,

I would prefer to keep the quality of images low in the jupyter notebooks while writing it as well as when storing it in git however when jupyter-book generates the pages I would prefer that the images are of high quality (retina). 

I generally have this at the top of my notebooks

```
%config InlineBackend.figure_format = "retina"
plt.rcParams["axes.spines.right"] = False
plt.rcParams["axes.spines.top"] = False
```

Is there a way to configure these types settings outside the notebook and provide a file (rc files, an additional section in _config.yml etc) to jupyter-book to use?

Regards & thanks
Kapil

**Comment by chrisjsewell** [comment]
Heya, IIRC no not currently I'm afraid.

I'm a little hesitant of modifying the actual contents of notebooks within jupyter-book.
But perhaps it would be interesting to see if environmental variables can be injected into the execution process, something like

```python
import os

if "JUPYTER_BOOK" in os.environ:
  %config InlineBackend.figure_format = "retina"
  plt.rcParams["axes.spines.right"] = False
  plt.rcParams["axes.spines.top"] = False
```

https://github.com/jupyter/nbclient/pull/79 may also be relevance

**Reply by choldgraf** [reply]
@chrisjsewell what if we allowed for something like a "prefix" to append to any page that is executed as a Jupyter Notebook? We could follow a pattern similar to the HTML sidebars pattern in Sphinx like:

```
ntbk_prefix = {
  "notebooks/*": """
      import matplotlib.pyplot as plt
      plt.rcparams(...
    """
}
```

and this would inject a cell w/ that code to be run before all other cells? It would need to be up to the author to know what kernel/language/etc would be used
?

**Reply by chrisjsewell** [reply]
I'm not really a fan of injecting code cells into the notebook, because now, if someone runs the notebook in e.g. JupyterLab, they might get a different result than from the one in the documentation, and it would not be clear why, as opposed to with the environmental variable approach where you can see what is not being run.

Injecting cells also may not play nicely with caching.

**Reply by choldgraf** [reply]
Isn't that just a slightly more generalized version of what you just suggested? (the if `"JUPYTER_BOOK" in os.environ` suggestion)

**Reply by chrisjsewell** [reply]
isn't what more generalized? adding code cells?

**Reply by choldgraf** [reply]
I thought you were suggesting adding a chunk of code that would be run before executing notebooks, and I thought I was just suggesting that we make this configurable on a `glob`-style pattern to only run on certain pages, but maybe I misunderstood your suggestion

**Reply by choldgraf** [reply]
ah ok I think I see what you're suggesting - so you'd let people run code, but *only* to set environmental variables that would then be available to the execution kernl, but no non-environment variables would be there

**Reply by chrisjsewell** [reply]
No, I was suggesting you literally add:

```python
import os

if "JUPYTER_BOOK" in os.environ:
  ...
```

directly to all notebooks where you want "divergent" behaviour.
then when executing notebooks, you make the environment include `JUPYTER_BOOK`

**Reply by chrisjsewell** [reply]
> so you'd let people run code, but only to set environmental variables

I wouldn't let people run anything, just have jupyter-cache inject this environmental variable into the environment in which the notebook is executed.

**Reply by chrisjsewell** [reply]
linking https://github.com/jupyter/nbclient/issues/115

**Reply by choldgraf** [reply]
oh interesting - got it

**Reply by mgeier** [reply]
Just for reference, in `nbsphinx` kernel arguments can be set in `conf.py` see https://nbsphinx.readthedocs.io/en/0.7.1/configuring-kernels.html#Kernel-Arguments

This can be used for Matplotlib settings, see https://nbsphinx.readthedocs.io/en/0.7.1/code-cells.html#Plots

The `ExecutePreprocessor` from `nbconvert` has the `extra_arguments` parameter for this.

**Reply by mgeier** [reply]
PS: I've asked a related question at https://github.com/jupyter/nbconvert/issues/1344, which I think would be a great feature, but this would *not* help in @ksachdeva's case (because those setting would also affect JupyterLab).

**Reply by chrisjsewell** [reply]
Cheers @mgeier, yeh we should most probably allow to parse such arguments also.
Although the environmental variable approach feels like it would be more generally applicable to all kernels/packages

**Reply by mgeier** [reply]
@chrisjsewell AFAICT, the `extra_arguments` parameter is generally applicable to all kernels/packages, isn't it?
The config option `nbsphinx_execute_arguments` is just a list of strings, kernels can interpret this however they see fit.

Nothing against environment variables though, those can of course also be used, see https://nbsphinx.readthedocs.io/en/0.7.1/configuring-kernels.html#Environment-Variables, but this creates some noise in the notebook.

**Comment by ksachdeva** [comment]
Thanks, guys; I think the usage of conditioning based on environment variable is an acceptable way to handle it. 

That said, I see the value of the external configuration as well. 

Instead of injecting code cells is there is a way to specify the directory/configuration/rc file for visualization libraries. Most of the visualization libraries have this notion of themes/config and incorporating these themes using the configuration options in jupyter-book may be another way to help customize the generation of final document format.

**Reply by loftusa** [reply]
Bumping on this. I'm writing a book at the moment and I'd like to have custom global `matplotlib` rcParams. Don't want to have to make a new cell at the top of every file...
